### ISSUE TITLE
support arm

### ISSUE THREAD

**Issue by tpoisonooo** [nbdd0121/stackful#1]
I notice that [fiber.s](https://github.com/nbdd0121/stackful/blob/master/src/fiber.s)  only works on intel x86.
Is there any possible to support `ARM` and  macos ?

**Comment by nbdd0121** [comment]
It certainly is possible. Unfortunately I don't have ARM or Mac development devices, so I couldn't develop and test myself.

**Comment by tpoisonooo** [comment]
I have bought an ARM64 AWS  instance, can you give me your email for sending the .pem ?

**Comment by nbdd0121** [comment]
You can clone the repo and use `git log` to get my email address.

**Comment by tpoisonooo** [comment]
the email has been sent, pls check.

**Comment by nbdd0121** [comment]
A preliminary AArch64 support is now available on the master branch. Feel free to have a try and report any issues.

Turns out it's much easier than I anticipated, although this is my first time writing any AArch64 assembly!

**Comment by tpoisonooo** [comment]
```bash
$ gcc --version
gcc (GCC) 8.3.1 20190311 (Red Hat 8.3.1-3)
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ uname -a
Linux 2102312YWM10M1000108 4.14.0-115.el7a.0.1.aarch64 #1 SMP Sun Nov 25 20:54:21 UTC 2018 aarch64 aarch64 aarch64 GNU/Linux
```

```bash
$ RUST_BACKTRACE=full ./run_with_plugins -p logical_test
thread 'blocking-1' panicked at 'failed to allocated stack', /home/security/.cargo/git/checkouts/stackful-322aae4d69ac55b8/ccbec34/src/lib.rs:109:17
thread 'stack backtrace:
main' panicked at 'failed to allocated stack', /home/security/.cargo/git/checkouts/stackful-322aae4d69ac55b8/ccbec34/src/lib.rs:109:17
   0:     0xaaaacd2a3168 - std::backtrace_rs::backtrace::libunwind::trace::hecc4fb3dc5a91037
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0xaaaacd2a3168 - std::backtrace_rs::backtrace::trace_unsynchronized::h9f238248263b1f81
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0xaaaacd2a3168 - std::sys_common::backtrace::_print_fmt::hae64a821d41b4b14
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:67:5
   3:     0xaaaacd2a3168 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha504e6c6649c19e9
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:46:22
   4:     0xaaaacd2c2e84 - core::fmt::write::h7c914d42a8e99f57
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/fmt/mod.rs:1110:17
   5:     0xaaaacd29da9c - std::io::Write::write_fmt::hb2f3ddc6101ed08d
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/io/mod.rs:1588:15
   6:     0xaaaacd2a51a8 - std::sys_common::backtrace::_print::h1e31b5a4bc5a5719
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:49:5
   7:     0xaaaacd2a51a8 - std::sys_common::backtrace::print::h3453f1f2e887cb14
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:36:9
   8:     0xaaaacd2a51a8 - std::panicking::default_hook::{{closure}}::hd0cacc29652b13c4
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:208:50
   9:     0xaaaacd2a4cc0 - std::panicking::default_hook::h0d7faf1abbe91b0d
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:225:9
  10:     0xaaaacd2a5800 - std::panicking::rust_panic_with_hook::he3ac366c10fca72f
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:622:17
  11:     0xaaaacd12def8 - std::panicking::begin_panic::{{closure}}::h0c750acb7ed196a8
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:542:9
  12:     0xaaaacd12e5ec - std::sys_common::backtrace::__rust_end_short_backtrace::hb62c6b6de75d4182
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:141:18
  13:     0xaaaacd12de7c - std::panicking::begin_panic::haee0888e537765b4
                               at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:541:12
  14:     0xaaaaccedd33c - stackful::Stack::allocate::h86aaf8361f483fcc
                               at /home/security/.cargo/git/checkouts/stackful-322aae4d69ac55b8/ccbec34/src/lib.rs:109:17
stack backtrace:
   Aborted (core dumped)
```

**Comment by nbdd0121** [comment]
Can you provide an example that allows me to replicate?

**Comment by nbdd0121** [comment]
Actually I might have a clue -- what's the page size of the system you are using? The code currently assumes 4K but that might not be true for all ARM systems.

**Comment by tpoisonooo** [comment]
> Actually I might have a clue -- what's the page size of the system you are using? The code currently assumes 4K but that might not be true for all ARM systems.

```bash
$ getconf PAGESIZE
65536
```

The code tested on Huawei Kunpeng Server (ARM64),  Huawei use "Huge Page" to accelerate applications.

**Comment by tpoisonooo** [comment]
> Can you provide an example that allows me to replicate?

The test code would be public tomorrow(20210910), now it's private.

**Comment by nbdd0121** [comment]
> ```shell
> $ getconf PAGESIZE
> 65536
> ```
> 
> The code test on Huawei Kunpeng Server (ARM64), Huawei use "Huge Page" to accelerate applications.

As expected. The master branch now already contains code that uses page size from sysconf instead of hard-coded 4096.

**Comment by tpoisonooo** [comment]
[MegFlow](https://github.com/MegEngine/MegFlow)  refers to `stackful`, now it's public.

**Comment by tpoisonooo** [comment]
thanks @nbdd0121 ,  MegFlow support aarch64.

**Comment by nbdd0121** [comment]
Published 0.1.1 on crates.io.
